# Day 2: ML Prediction
### We use supervised regression techniques to predict cycle times using a certain number of features

In [19]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

#Setting a random seed so that random numbers are repeatable
np.random.seed(42)

#no of datapoints
n = 250

# Synthetic operational variables
empty_distance_km = np.random.uniform(1, 6, n)
loaded_distance_km = np.random.uniform(1, 6, n)
queue_length = np.random.randint(0, 6, n)
truck_age_years = np.random.uniform(1, 12, n)

# Synthetic target: cycle time in minutes
cycle_time = (
    8
    + 2.2 * empty_distance_km
    + 3.0 * loaded_distance_km
    + 4.5 * queue_length
    + 0.5 * truck_age_years
    + np.random.normal(0, 3, n)
)

#creating a dataframe with the synthetic data
df = pd.DataFrame({
    "empty_distance_km": empty_distance_km,
    "loaded_distance_km": loaded_distance_km,
    "queue_length": queue_length,
    "truck_age_years": truck_age_years,
    "cycle_time": cycle_time
})

#assigning Features and target variable
X = df[["empty_distance_km", "loaded_distance_km", "queue_length", "truck_age_years"]]
y = df["cycle_time"]


# Splitting the data in train-test with 3:1 ratio
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)


#initialising model object dictionary from skikit library.
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42)
}
#looping through the dictionary for both models
for name, model in models.items():
    model.fit(X_train, y_train) #trains the model with the training features and target value
    
    pred = model.predict(X_test) #predicting target values for test features

    print(name) #model name 
    print(" MAE:", round(mean_absolute_error(y_test, pred), 2)) #printing mean absolute error
    print(" R2 :", round(r2_score(y_test, pred), 3)) #printing r2 score

    
print("Mean cycle time:", y.mean())

Linear Regression
 MAE: 2.56
 R2 : 0.893
Random Forest
 MAE: 3.1
 R2 : 0.844
Mean cycle time: 40.87788496640907


# How to decide which model is better?
We have MAE and R2 score for each model. Lets understand what these mean and how do they help

## MAE (Mean Absolute Error)

Measures the average prediction error between actual and predicted values.

$$
MAE = \frac{1}{n} \sum |y_i - \hat{y}_i|
$$

---

## R² Score

Measures how well the model explains the variation in data.

$$
R^2 = 1 - \frac{\sum (y_i - \hat{y}_i)^2}{\sum (y_i - \bar{y})^2}
$$

Since, MAE is relative, we need to check the order of target variable. y.head() gives us a value around 40. an MAE of 2-3 is fine for this range
Because MAE is just an error value, the lesser the MAE, the accurate the prediction is. 

For R2 score, it shows us that how well hte variation in the data is explained by the model. 
R2 = 1 means a perfect model, i.e., it predicts every datapoint correctly. 
R2 = 0 means the model is no better than average value.
R2 < 0 means that the model is worse than average results.

### Conclusion
Comparing the MAE and R2 for both models, we can conclude that Linear Regression works better in this scenario. 

Since the target variable is essentially a linear combination of all the features, the linear regression model works best here. 
On the other hand, if it had been a non-linear combination, Random forest would have outperformed Linear regression.

---
# Modification with new feature
___
In the following section, we add a new feature slope_percent. The cycle time increases with slope_percent. 

In [18]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

#Setting a random seed so that random numbers are repeatable
np.random.seed(42)

#no of datapoints
n = 250

# Synthetic operational variables
empty_distance_km = np.random.uniform(1, 6, n)
loaded_distance_km = np.random.uniform(1, 6, n)
queue_length = np.random.randint(0, 6, n)
truck_age_years = np.random.uniform(1, 12, n)

#adding the new feature, integers from 0 to 100
slope_percent = np.random.randint(0, 101, n)

# Synthetic target: cycle time in minutes
cycle_time = (
    8
    + 2.2 * empty_distance_km
    + 3.0 * loaded_distance_km
    + 4.5 * queue_length
    + 0.5 * truck_age_years
    #having the cycle time increase with slope percent. Multiplied by 0.2 to keep weights equivalent
    + 0.2 * slope_percent
    + np.random.normal(0, 3, n)
)

#creating a dataframe with the synthetic data
df = pd.DataFrame({
    "empty_distance_km": empty_distance_km,
    "loaded_distance_km": loaded_distance_km,
    "queue_length": queue_length,
    "truck_age_years": truck_age_years,
    # Adding new feature in dictionary
    "slope_percent" : slope_percent,
    "cycle_time": cycle_time
})

#assigning Features and target variable
# slope percent assigned to feature series
X = df[["empty_distance_km", "loaded_distance_km", "queue_length", "truck_age_years", "slope_percent"]]
y = df["cycle_time"]


# Splitting the data in train-test with 3:1 ratio
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)


#initialising model object dictionary from skikit library.
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42)
}
#looping through the dictionary for both models
for name, model in models.items():
    model.fit(X_train, y_train) #trains the model with the training features and target value
    
    pred = model.predict(X_test) #predicting target values for test features

    print(name) #model name 
    print(" MAE:", round(mean_absolute_error(y_test, pred), 2)) #printing mean absolute error
    print(" R2 :", round(r2_score(y_test, pred), 3)) #printing r2 score

    
print("Mean Cycle time:", y.mean())

Linear Regression
 MAE: 2.57
 R2 : 0.921
Random Forest
 MAE: 3.55
 R2 : 0.861
Mean Cycle time: 50.92501086905966


---
# Results
On Adding a new feature _slope_percent_, the overall prediction does improve by some value. However, this is quite dependent on the scale or weight
of the feature in generating the synthetic target variable. 

# Model Performance Comparison

## Linear Regression

| Metric | Earlier | Now |
|---|---|---|
| %MAE | 6.26% | 5.05% |
| R² Score | 0.893 | 0.921 |

---

## Random Forest

| Metric | Earlier | Now |
|---|---|---|
| %MAE | 7.58% | 6.97% |
| R² Score | 0.844 | 0.861 |

---

### Observation

- Both models improved after adding the new feature.
- Linear Regression still performs better.
- The target relationship remains mostly linear.

_note_: %MAE is shown because average cycle time increases by introducing slope percent